sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/dependency.py

```diff
@@ -331,6 +350,313 @@ def get_dependencies(node, dependency_type, relationship):
     return dependencies
 
 
+def filter_primary_science_dependencies(
+    session: db.Session, records: list, query_data_type: str
+):
+    """Filter primary science dependencies for unprocessed downstream dependencies.
+
+    Parameters
+    ----------
+    session : orm session
+        Database session.
+    records : list
+        Science file records.
+    query_data_type : str
+        The query data type.
+
+    Returns
+    -------
+    list
+        Upstream primary source files that have downstream dependencies that need
+        to be processed.
+    """
+    files = []
+    for record in records:
+        # check in the science table if the upstream primary source dependency
+        # already exists.
+        query = select(models.ScienceFiles.__table__).where(
+            models.ScienceFiles.instrument == record.instrument,
+            models.ScienceFiles.descriptor == record.descriptor,
+            # Use the query data type instead of the current record.
+            models.ScienceFiles.data_level == query_data_type,
+            models.ScienceFiles.start_date == record.start_date,
+            models.ScienceFiles.version == record.version,
+        )
+        # If the upstream primary source dependency does not exist, add it to the list
+        # of files to return.
+        # This indicates that the upstream primary source needs to be processed.
+        upstream_primary_source = session.execute(query).first()
+
+        if not upstream_primary_source:
+            files.append(basename(record.file_path))
+
+    return files
+
+
+def primary_science_dep(query: dict, dependency: dict) -> bool:
+    """Check if the dependency is a primary science dependency.
+
+    A primary science dependency exists when both the upstream and downstream
+    dependencies have the same data source and both are valid science data types.
+
+    Parameters
+    ----------
+    query : dict
+        The dependency from the api call.
+    dependency : dict
+       Upstream or downstream dependency from the query.
+
+    Returns
+    -------
+    bool
+        True if dependency is a primary science dependency, False otherwise
+
+    Examples
+    --------
+    swe_l1b_sci is a primary science dependency of swe_l1a_sci.
+    """
+    return (
+        query["data_source"] == dependency["data_source"]
+        and valid_science(dependency["data_type"])
+        and valid_science(query["data_type"])
+    )
+
+
+def get_dependency_processing_input(
+    query: dict,
```

Comment:
can you change this to `query_params` like you input names?
```suggestion
    query_params: dict,
```